crates/swc_ecma_utils/src/function/function_wrapper.rs

```diff
@@ -0,0 +1,422 @@
+                return_type: Default::default(),
+            },
+        };
+
+        (fn_expr, ref_decl)
+    }
+
+    ///
+    /// ```javascript
+    /// function NAME(PARAMS) {
+    ///     return REF.apply(this, arguments);
+    /// }
+    /// ```
+    fn build_function_forward(&mut self, ref_ident: Ident, name_ident: Option<Ident>) -> FnExpr {
+        let apply = Stmt::Return(ReturnStmt {
+            span: DUMMY_SP,
+            arg: Some(Box::new(ref_ident.apply(
+                DUMMY_SP,
+                Box::new(Expr::This(ThisExpr { span: DUMMY_SP })),
+                vec![quote_ident!("arguments").as_arg()],
+            ))),
+        });
+
+        FnExpr {
+            ident: name_ident,
+            function: Function {
+                span: DUMMY_SP,
+                is_async: false,
+                is_generator: false,
+                params: self.params.take(),
+                body: Some(BlockStmt {
+                    span: DUMMY_SP,
+                    stmts: vec![apply],
+                }),
+                decorators: Default::default(),
+                type_params: Default::default(),
+                return_type: Default::default(),
+            },
+        }
+    }
+}
+
+impl From<FnExpr> for FunctionWrapper<Expr> {
+    fn from(mut fn_expr: FnExpr) -> Self {
+        let function_ident = fn_expr.ident.take();
+        let params = Self::get_params(fn_expr.function.params.iter());
+        Self {
+            binding_ident: None,
+            function_ident,
+            params,
+            function: fn_expr.into(),
+            _type: Default::default(),
+        }
+    }
+}
+
+impl From<ArrowExpr> for FunctionWrapper<Expr> {
+    fn from(
+        ArrowExpr {
+            span,
+            params,
+            body,
+            is_async,
+            is_generator,
+            ..
+        }: ArrowExpr,
+    ) -> Self {
+        let body = Some(match body {
+            BlockStmtOrExpr::BlockStmt(block) => block,
+            BlockStmtOrExpr::Expr(expr) => BlockStmt {
+                span: DUMMY_SP,
+                stmts: vec![Stmt::Return(ReturnStmt {
+                    span: expr.span(),
+                    arg: Some(expr),
+                })],
+            },
+        });
+
+        let function = Function {
+            span,
+            params: params.into_iter().map(Into::into).collect(),
+            decorators: Default::default(),
+            body,
+            type_params: None,
+            return_type: None,
+            is_generator,
+            is_async,
+        };
+
+        let fn_expr = FnExpr {
+            ident: None,
+            function,
+        };
+
+        Self {
+            binding_ident: None,
+            function_ident: None,
+            params: Self::get_params(fn_expr.function.params.iter()),
+            function: fn_expr.into(),
+            _type: Default::default(),
+        }
+    }
+}
+
+impl Into<Expr> for FunctionWrapper<Expr> {
```

Comment:
Oh, I didn't know this impl is possible.

I'm not sure why does it work. It should be one of `specialization` or `#[fundameltal]`.